reconcile/vpc_peerings_validator.py

```diff
@@ -41,13 +42,59 @@ def validate_no_internal_to_public_peerings(
                 continue
 
             valid = False
+            pair = {cluster.name, peer.name}
+            if pair in found_pairs:
+                continue
+            found_pairs.append(pair)
             logging.error(
                 f"found internal to public vpc peering: {cluster.name} <-> {peer.name}"
             )
 
     return valid
 
 
+def validate_no_public_to_public_peerings(
+    query_data: VpcPeeringsValidatorQueryData,
+) -> bool:
+    """Iterate over VPC peerings of public clusters and validate the peer is not public."""
+    valid = True
+    found_pairs: list[set[str]] = []
+    for cluster in query_data.clusters or []:
```

Comment:
i'm just copying from @fishi0x01: https://github.com/app-sre/qontract-reconcile/blob/d3620c7029641438f6e8c73529ae160b7a51ccd7/reconcile/service_dependencies.py#L79

Reply:
@steveteahan this is a fair point. Im wondering if it is possible to address this concern in general in the gql schema itself. All our queries are defined to return lists. To me it doesn't really makes sense that those are always nullable, instead of empty lists if there is no such data. I wonder if this is a GQL limitation or if we could easily adjust the schema there.

Reply:
I.e., in general (not only query), if a type has `isList: true` set in our schema, should it really be nullable? Shouldn't it always return an empty list at least?

Reply:
We'd probably need to adjust the gql server https://github.com/app-sre/qontract-server/blob/master/src/schema.ts#L196-L198

Will play around with this a little.

Reply:
> Im wondering if it is possible to address this concern in general in the gql schema itself. All our queries are defined to return lists. To me it doesn't really makes sense that those are always nullable, instead of empty lists if there is no such data. I wonder if this is a GQL limitation or if we could easily adjust the schema there.

I'm 100% with you on the idea that if it's `isList` then it'd make more sense to me to return an empty list instead of `None`. I was actually trying to find more information to see what the expected behavior is in these situations for GraphQL. My concern with changing it is whether it could be a breaking API change if `None` was being checked directly before.

I think the general point stands that in lieu of some improvements to greatly reduce `Optional` everywhere, we can always check for cases where the list is empty and filter only those resources matching the desired attributes to avoid the need to `None` check everywhere.

In this case, it could be something like (I didn't actually test this):

```python
if not query_data.clusters:
    <exit integration immediately>

clusters = [cluster for cluster in query_data.clusters if cluster.peering.connections]
```

In theory, that'd remove the possibility for `Optional` as long as Mypy is smart enough to know that we checked for `None` already.